crates/sui-faucet/src/faucet/simple_faucet.rs

```diff
@@ -548,6 +556,129 @@ impl SimpleFaucet {
         Ok((res.digest, coin_ids))
     }
 
+    async fn build_batch_pay_sui_txn(
+        &self,
+        coin_id: ObjectID,
+        batch_requests: Vec<(Uuid, SuiAddress, Vec<u64>)>,
+        signer: SuiAddress,
+        budget: u64,
+    ) -> Result<TransactionData, anyhow::Error> {
+        let gas_payment = self.wallet.get_object_ref(coin_id).await?;
+        let gas_price = self.wallet.get_reference_gas_price().await?;
+        let pt = {
+            let mut builder = ProgrammableTransactionBuilder::new();
+            for (_uuid, recipient, amounts) in batch_requests {
+                let recipients: Vec<SuiAddress> =
+                    std::iter::repeat(recipient).take(amounts.len()).collect();
+                builder.pay_sui(recipients, amounts)?;
+            }
+            builder.finish()
+        };
+
+        Ok(TransactionData::new_programmable(
+            signer,
+            vec![gas_payment],
+            pt,
+            budget,
+            gas_price,
+        ))
+    }
+    // Function to process the batch send of the mcsp queue
+    pub async fn batch_transfer_gases(&self) -> Result<TransactionDigest, FaucetError> {
+        let Ok(mut req_consumer) = tokio::time::timeout(LOCK_TIMEOUT, self.request_consumer.lock()).await else {
+                error!("Timeout when getting request consumer lock");
+                return Err(FaucetError::Internal("Unable to get request queue lock.".to_string()));
+            };
+
+        let mut counter = 0;
+        let mut requests = Vec::new();
+        // We take the first 500 items off the queue or when it runs out.
+        while counter < 500 {
+            match timeout(RECV_TIMEOUT, req_consumer.recv()).await {
```

Comment:
The way this is currently architected, you can, in the worst case, end up waiting `RECV_TIMEOUT * 500` for your request to go through, which seems bad.  We should set an overall timeout for the process of gathering requests, at which point we create a batch out of whatever requests we have and send that out.

Reply:
Left a reply on the existing thread about this, but it's on an out-of-date change, so duplicating it here:

> I'm not 100% on this, but I believe a timer with a loop may not work as intended, because the tokio runtime does not know to wake up your task when it times out in that case, whereas with a call to `timeout` it will. The way I would try to do this with timeout is to pass in the vector to accumulate the batch as a `&mut` param to the function.